### PR TITLE
Ruby On Rails Recipe App Project

### PR DESCRIPTION
**Pull request by NatanimA**
[Presentation link](https://drive.google.com/file/d/1wF5hUUHQ5Jxadt6z_gJs7FULn_qI0zK8/view)
### Project requirements

- You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
- Login page and registration page:
    - Should be built with Devise.
- Food list:
    - Should display a list of food added by the logged-in user as in the wireframe (for teams with only 2 people - display also quantity of a given food).
    - Should lead to a form that allows users to add new food.
- Recipes list:
    - Should display a list of recipes created by the logged-in user as in the wireframe.
    - Should lead to recipe details.
    - If the user is the owner of the recipe, should allow the user to delete it.
- Public recipe list:
    - Should display a list of all public recipes ordered by newest as in the wireframe.
    - Should lead to recipe details.
- Recipe details:
    - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
    - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
- General shopping list view (**only for a group that has less than 3 members**):
    - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    - Should count the total food items and total price of the missing food.
- Make sure there are no N+1 queries happening.
- Create a navigation menu that allows users to open all of the pages you created.
---
### TO RUN THE APPLICATION PLEASE FOLLOW THE FOLLOWING

- [ ] Configure your Database
- [ ] Install all dependencies ```bundle install```
- [ ] Migrate the schemas ```rails db:migrate```
- [ ] To test is `rspec`

blog_app_test

- [ ] Make sure you have data to be rendered in your database
- [ ] Run ```rails s```
- [ ] Make sure to sign up 
- [ ] check your email for validations
- [ ] login
- [ ] Enjoy 😉

### CODE REVIEW

File: app/controllers/foods_controller.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        else
+          flash[:alert] = @recipe_foods.errors.full_messages.join(', ')
+        end
+        redirect_to recipe_path(id: recipe_id)
+      end
+    else
+      flash[:alert] = @food.errors.full_messages.join(', ')
+      food = Food.new
+      # format.html { redirect_to request.referrer, locals: { food: food } }
+      redirect_to new_food_path, locals: { food: }
```
Comment (Deepakdanger) [inline app/controllers/foods_controller.rb]: - [x] I've got an error when I start the page. I'm not sure if you got an error or not, but when I add there 

redirect_to new_food_path, locals: { food: **@food** }, then it fixed.

Screenshot from 2023-02-10 21-34-33

Please fix that.

File: app/controllers/foods_controller.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          flash[:notice] = 'Ingredient created successfully!'
+        else
+          flash[:alert] = @recipe_foods.errors.full_messages.join(', ')
+        end
+        redirect_to recipe_path(id: recipe_id)
+      end
+    else
+      flash[:alert] = @food.errors.full_messages.join(', ')
+      food = Food.new
+      # format.html { redirect_to request.referrer, locals: { food: food } }
```
Comment (Deepakdanger) [inline app/controllers/foods_controller.rb]: - [x] I would recommend removing the commented code to make the codebase cleaner.

File: app/controllers/foods_controller.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
+class FoodsController < ApplicationController
+  load_and_authorize_resource
+  before_action :set_food, only: %i[show destroy]
+  # before_action :authorize_delete, only: [:destroy]
```
Comment (Deepakdanger) [inline app/controllers/foods_controller.rb]: - [x] Here as well, I would recommend removing the commented code to make the codebase cleaner.

File: app/controllers/recipes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  def public
+    @recipes = Recipe.where(public: true).includes(:recipe_foods)
+  end
+
+  def new
+    recipe = Recipe.new
+    respond_to do |format|
+      format.html { render :new, locals: { recipe: } }
+    end
```
Comment (Deepakdanger) [inline app/controllers/recipes_controller.rb]: - [x] On line no. 16 and 51. I'm getting the error. Please fix that.

format.html { render :new, locals: { recipe: **recipe**} }


Screenshot from 2023-02-10 21-37-28

File: db/migrate/20230206112201_change_qunatity_from_foods.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class ChangeQunatityFromFoods < ActiveRecord::Migration[7.0]
+  def change
+    rename_column :foods, :qunatity, :quantity
+    #Ex:- rename_column("admin_users", "pasword","hashed_pasword")
```
Comment (Deepakdanger) [inline db/migrate/20230206112201_change_qunatity_from_foods.rb]: - [ ] I would recommend removing the commented code to make the codebase cleaner.

**Review by Deepakdanger — CHANGES_REQUESTED** [review]
Hi @NatanimA & @j-tee, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## To Highlight :100: 
- No linter error :heavy_check_mark: 
- Website looks professional :+1: 

## Required changes
_Check the comments under the review._


## Optional changes
You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by ShadyShawkat — APPROVED** [review]
## Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- All linters are passing ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**